prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"crypto/tls"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// A fetcher for a GCS client
+type GCSArtifactFetcher struct {
+	client      *storage.Client
+	xmlEndpoint string
+	withTLS     bool
+}
+
+// A location in GCS where Prow job-specific artifacts are stored. This implementation assumes
+// Prow's native GCS upload format (treating GCS keys as a directory structure), and is not
+// intended to support arbitrary GCS bucket upload formats.
+type GCSJobSource struct {
+	source     string
+	linkPrefix string
+	bucket     string
+	jobPath    string
+	jobName    string
+	buildID    string
+}
+
+// NewGCSArtifactFetcher creates a new ArtifactFetcher with a real GCS Client
+func NewGCSArtifactFetcher(c *storage.Client, xmlEndpoint string, tls bool) *GCSArtifactFetcher {
+	return &GCSArtifactFetcher{
+		client:      c,
+		xmlEndpoint: xmlEndpoint,
+		withTLS:     tls,
+	}
+}
+
+// NewGCSJobSource creates a new GCSJobSource from a given bucket and jobPath
+func NewGCSJobSource(src string) *GCSJobSource {
+	linkPrefix := "gs://"
```

Comment:
Can you isolate this logic to a `parseGcsURI` or something  so we can have concise unit tests for it?

Reply:
Steve don't you and I both have instances of this running around somewhere? Here's mine, feel free tomove out of testgrid

https://github.com/kubernetes/test-infra/blob/7c80a99067aebd61c99e16e8ee8b88a435390fd0/testgrid/util/gcs/gcs.go#L46-L84

Reply:
oh thats excellent, ill definitely be using that

Reply:
Yeah I think I linked my version last time, too :) It's somewhere under pod-utils/gcs